<br>
<br>
<br>
<br>

# DAV 6150 Module 7: Regression Modeling for Categorical  Response Variables
<br>
<br>
<br>

## Module 5 Assignment Review

### A Simple Method for Calculating a ROC Curve

A ROC curve is formulated by plotting the __true positive rate__ (TPR) (a.k.a., __sensitivity__) against the __false positive rate__ (FPR) (a.k.a, __1 - specificity)__ for a given classifier's actual classifications and scored probabilities.

The FPR is represented by the X axis while the TPR is represented by the Y axis.

__Remember__:  

- The __true positive rate__ (TPR) is the proportion of actual "positive" classification values that are __correctly identified as "positive"__. This is also known as the __sensitivity__ of a classifier, i.e., TP / (TP + FN)


- The __false positive rate__ (FPR) is the proportion of actual "negative" classification values that are __incorrectly identified as "positive"__. FPR is defined as FP / (FP + TN), which is the same as __1 - specificity__ (remember that specificity = TN/(TN + FP) )


To derive the values required for a ROC plot of a model we __iterate through an ordered sequence of 'K' classification threshold values__ within the range of $(0 < threshold <= 1)$ (e.g., if 'K' == 100, our iterator would span .01 through 1.00 via .01 increments).

While iterating through the threshold values, we __compare the scored probabilities against each iterative threshold value__ to derive + store TPR and FPR values across the full range of iterable threshold values. When finished, we have accumulated 'K' pairs of TPR and FPR values.

Then, plot the resulting TPR and FPR metrics against one another on a 2-dimensional plane.

#### A Python-based example

In [4]:
''' PLEASE NOTE THAT THIS PYTHON CODE IS SHOWN ONLY FOR PURPOSES OF EXPLAINING
HOW TO APPROACH THE CREATION OF A ROC CURVE FOR A BINARY CLASSIFIER
TO MAKE USE OF IT YOU WOULD NEED TO INCORPORATE IT WITHIN YOUR OWN CONTEXT

'actual' is the class variable (the actual classification) from the data set
'probability' is the scored.probability variable from the data set '''

# -----------------------------------------------------------------------
import numpy as np
#create list of thresholds
thresholds = np.arange(0,1,0.01)

#create list of true positive rates
tpr_list=[]

  #create list of false positive rates
fpr_list=[]

#create new list of predictions for every threshold
for i in thresholds: #for every threshold
  new_predicted=[] #create new list of predictions
  for j in range(len(probability)): #for each observation
  #compare threshold to probability and append new prediction
    if probability[j] < i:
      new_predicted.append(0)
    else:
      new_predicted.append(1)

  #calculate the tpr
  tpr = sensitivity(actual, pd.Series(new_predicted))
  tpr_list.append(tpr) #append the true positive rate to the list

  #calculate the fpr
  fpr = 1 - specificity(actual, pd.Series(new_predicted))
  fpr_list.append(fpr) #append the false positive rate to the list

# ...then continue to formulate your code for plotting tpr_list vs. fpr_list
# via the Python graphics tool of your choice

# Module 7: Regression Modeling for Categorical Response Variables

- __Binary Logistic Regression__: The response variable is a __binary categorical variable__, while the explanatory variables can be either continuous, discrete or binary (e.g., dummy variables created from the values of a categorical variable).


- __Multinomial Logistic Regression (MLR)__: The response variable is a categorical variable having __more than two (2) possible values__, while the explanatory variables can be either continuous, discrete or binary. There are many types of MLR models, and the exact type of MLR model you choose to use will be dependent on whether your response variable is __ordinal__ or __nominal__ in nature.


### Binary Logistic Regression

In binary logistic regression, the response variable is __binary__, i.e., it contains data encoded as either '1' (e.g., 'True') or '0' (e.g., 'False'). In other words, logistic regression is used for __classification__ problems.

Logistic regression finds the best fitting __mathematical model__ (not a simple linear equation) to describe the relationship between the binary response variable and the explanatory variable(s). The values it generates are the coefficients of a formula to predict a __logit transformation__ (aka "__log odds"__) which is the logarithm of the odds $p$ of the probability of the presence of the characteristic of interest.

$logit(p) = ln(p/(1-p)) = b0 + b1x1 + b2x2 + .. + bnxn$

where $p$ = the probability of the __presence__ of a characteristic,

$(1-p)$ = the probability of the __abscence__ of a characteristic,

$b0$ = is a constant value

$b1, b2, ..,bn$ = the regression coefficients for the explanatory variables $x1, x2, .., xn$

Effective binary logistic regression models contain little to no collinearity amongst the explanatory variables used.

The explanatory variables should be linearly related to the log odds of the probability of the presence of the characteristic of interest (i.e., the value of the response variable)

Logistic regression generally does not work well when applied to very small data sets. For observational studies, a minimum of __500 samples/observations__ is recommended.

Further explanation is available here:

https://www.statisticssolutions.com/what-is-logistic-regression/



### Multinomial Logistic Regression

In multinomial logistic regression (MLR), the response variable is __multiclass__, i.e., it has __more than 2 possible values__. Multinomial regression calculates the probability of any observation being assigned to each of the possible classes. Therefore, the regression model must perform at least $n-1$ calculations, where $n$ is the number of possible classifications for the response variable.

Why $n-1$ calculations? If we have $n$ classes, we can determine the probability of the $n$th class by simply summing the probabilities for each of the other classifications and subtracting that sum from $1$, i.e.,

### $P(Class_n) = 1 - ( P(Class_1) + P(Class_2) + ... + P(Class_{n-1}) ) $

As with binary logistic regression, the coefficients of a MLR model predict the __log odds__ of the probability of any observation belonging to each of the $n$ classes.

The probability of observing class $k$ out of $n$ total classes is:

# $ P(y_i = k) = \frac { e^{S_k} }  {e^{S_1} + e^{S_2} + ... + e^{S_n} } $

(see https://towardsdatascience.com/understanding-logistic-regression-coefficients-7a719ebebd35)

When performing multiple logistic regression, the algorithm assigns a predicted classification for a given observation based on __which of the possible classifications achieves the largest log odds value__.


### How to interpret logistic regression model coefficients

"A one unit increase in explanatory variable $x$ (with all other explanatory variables held constant) increases the log-odds of response variable $y$ by the amount indicated by the coefficient of $x$."

However, in practice it is very common to simply examine the __directionality__ of the coefficients, i.e., if a coefficient is __negative__, then the larger the value of the explanatory variable, the more it will tend to __decrease__ the magnitude of the response variable. Conversely, if a coefficient is __positive__, then the larger the value of the explanatory variable, the more it will tend to __increase__ the magnitude of the response variable.


### How to convert log odds to a probability

- __Step 1__: Convert the log odds value to odds via __exponentiation__, i.e., the antilog of a log value. Assuming the log odds value we've been given is a __natural log__, we apply an __exponential function__ to convert the log odds value to and odds value.


- __Step 2__: Calculate the probability as follows:

### $ Prob = \frac {odds} {1 + odds} $

http://www.pmean.com/13/predicted.html


### Advantages of Logistic Regression

- Relatively computationally efficient + easy to implement


- Relatively easy to interpret (though not as easy to interpret as linear regression models)


- Can be effective even if features have not been scaled to similar ranges


- Due to its simplicity, logistic regression is widely used as a baseline when comparing the performance of other more complex classification methods (e.g., decision trees, random forest, KNN, neural networks, etc.)


### Disadvantages of Logistic Regression

- Due to its simplicity, performance can lag behind that of more complex classification methods


- Requires a __linear__ relationship between the explanatory variables and the response variable since it relies on a linear decision space


- Requires that explanatory variables are __very__ independent of one another + that they have a substantive relationship with the response variable.

### Evaluating Logistic Regression Model Performance

In __Module 5__ we learned about a wide variety of classification model performance metrics, all of which can be used for purposes of evaluating the performance of any type of logistic regression model:

- Confusion Matrices: True Positives (TP), False Positives (FP), True Negatives (TN), False Negatives
 (FN)


- Accuracy


- Precision


- Recall / Sensitivity


- Specificity


- F1 Score


- ROC: __ROC (Receiver Operating Characteristic) Curve__: Calculate by plotting the __true positive rate (TPR)__ against the __false positive rate (FPR)__; http://www.saedsayad.com/model_evaluation_c.htm. Plotting TPR vs. FPR for a series of classification models constructed using different thresholds for predicting classifications yields a curve within a two-dimensional plane.


- AUC: In a ROC plot, AUC is determined by calculating the area in the plot that falls __below / to the right of__ the ROC curve.  A random classifier has an area under the curve of 0.5, while AUC for a perfect classifier is equal to 1. The higher the AUC score, the better the performance of a model.  AUC scores from different models can be compared against one another to help us determine which model has the best performance.

## Logistic Regression example with scikit-learn

From the "Python For Data Analysis" textbook (from your DAV 5400 course): Use the 'Titanic' data set which contains passenger survival rate data for the ocean-going vessel of that name that sank in 1912.

Our goal is to __try to predict whether or not a passenger would have been more likely than not to survive the sinking of the Titanic__ based on the data contained in the data set (i.e., build a __predictive model__).

- First, we will fit a logistic regression model on a __training__ data set using the data set's __PClass__, __Age__, and __Sex__ attributes.


- Then we evaluate the model using a __testing__ data set that is exclusive of the data contained in the __training__ data set.


In [ ]:
# load the LogisticRegression() function from sklearn's 'linear_model' sub-library
from sklearn.linear_model import LogisticRegression

# load the training + testing data sets for the Titanic data
train = pd.read_csv('https://raw.githubusercontent.com/wesm/pydata-book/2nd-edition/datasets/titanic/train.csv')
test = pd.read_csv('https://raw.githubusercontent.com/wesm/pydata-book/2nd-edition/datasets/titanic/test.csv')

# display the first four rows of the data set
train[:4]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S

In [ ]:
# lets check the first four rows of the test data
# Note the lack of a "Survived" indicator !!!
test[:4]

PassengerId  Pclass                              Name     Sex   Age  SibSp  \
0          892       3                  Kelly, Mr. James    male  34.5      0   
1          893       3  Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
2          894       2         Myles, Mr. Thomas Francis    male  62.0      0   
3          895       3                  Wirz, Mr. Albert    male  27.0      0   

   Parch  Ticket    Fare Cabin Embarked  IsFemale  
0      0  330911  7.8292   NaN        Q         0  
1      0  363272  7.0000   NaN        S         1  
2      0  240276  9.6875   NaN        Q         0  
3      0  315154  8.6625   NaN        S         0

In [ ]:
# how many records in the training data set?
train.shape

(891, 12)

In [ ]:
# how many people survived in the training set?
train.Survived.values.sum()

342

In [ ]:
# what percentage of the training set survived?
train.Survived.values.sum() / train.shape[0]

0.3838383838383838

__NOTE__: Since we know that 38.3% of the people in the training set survived, we could achieve a training model accuracy of (1 - .383) = 61.7% by simply predicting "Did not survive" for each passenger. This metric is referred to as the __null error rate__.  When evaluating the performance of a binary logistic regression model, always check to see whether the accuracy you are attaining exceeds the __null error rate__. If not, your model is unlikely to be of any value.


Missing data will generally prevent the fitting of a model via either __scikit-learn__ or __statsmodels__, so we must first check both the training and testing data for missing data values:

In [ ]:
# check the training data for null values
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Within the training data, the __Age__, __Cabin__ and __Embarked__ attributes all have missing data.

In [ ]:
# check the test data for null values
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

Within the testing data, the __Age__, __Fare__ and __Cabin__  attributes all have missing data.

So if we want to use __Age__ as part of our predictive model, we need to somehow fill the missing __Age__ values. In this instance, the author of the PfDA text chooses to use the relatively simple process of filling the missing values with the __median__ Age value found within the __training__ data set:

In [ ]:
# find the median 'Age' value within the training data set
impute_value = train['Age'].median()

# now fill the missing 'Age' values in both the training and testing data sets
train['Age'] = train['Age'].fillna(impute_value)
test['Age'] = test['Age'].fillna(impute_value)

Next, __we create a dummy indicator for the 'Sex' categorical variable__: the new dummy variable 'IsFemale' contains a '1' if the 'Sex' value for a passenger is 'Female' and a '0' otherwise:

In [ ]:
# create a dummy variable for the 'Sex' attribute in both the training and
# testing data sets
train['IsFemale'] = (train['Sex'] == 'female').astype(int)
test['IsFemale'] = (test['Sex'] == 'female').astype(int)

Now we are ready to define our predictive model using the attributes we want to use as explanatory variables:

In [ ]:
# define a vector containing the names of the attributes to use
predictors = ['Pclass', 'IsFemale', 'Age']

# create a subset of the training data using ONLY the selected explanatory variables
X_train = train[predictors].values

# create a subset of the testing data using ONLY the selected explanatory variables
X_test = test[predictors].values

# isolate the response indicator for the training data
y_train = train['Survived'].values

# sanity check on training data
X_train[:5]

array([[ 3.,  0., 22.],
       [ 1.,  1., 38.],
       [ 3.,  1., 26.],
       [ 1.,  1., 35.],
       [ 3.,  0., 35.]])

In [ ]:
# sanity check on response indicator
y_train[:5]

array([0, 1, 1, 1, 0], dtype=int64)

In [ ]:
# We're using the LogisticRegression() method for this model
model = LogisticRegression()

# fit the model: X_train contains our explanatory variables while
# y_train contains the response variable
model.fit(X_train, y_train)

C:\Users\James T\Anaconda3\envs\DAV5400\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [ ]:
# calculate the accuracy of the model relative to the training data set
model.score(X_train, y_train)

0.7946127946127947

Recall from above we calculated the __null error rate__ for our data to be 61.7%. The model we've generated has an accuracy score of 79.46%. As such, our model appears to be useful.

__What can we learn by examining the regression model coefficients for the explanatory variables?__

While the actual numeric values of the coefficients of a regression model are often very difficult to interpret, we can use the fact that a coefficient is either positive or negative to determine what effect of an __increase__ in the value of a given variable will have on the predicted value of the response variable.

In this binary logistic regression example, the value of the response variable is the "log odds" of the binary outcome being either a '0' or '1' ('0' meaning the passenger was more likely to perish while '1' indicates the passenger was more likely to survive).

In [ ]:
# examine the model coefficients for the explanatory variables
print(predictors)
model.coef_

['Pclass', 'IsFemale', 'Age']


array([[-1.07373582,  2.52093733, -0.02864864]])

From the above we see that:

- __Passenger Class__: An increase in the value of 'Passenger_Class' is associated with a __decreased__ likelihood of survival, i.e., passengers in lower classes of service were more likely to perish than were passengers in relatively higher classes of service.


- __IsFemale__: Being female increased the likelihood of survival


- __Age__: The older the passenger, the less likely they were to survive.

In [ ]:
# generate predictions for the test data using our new model
y_predict = model.predict(X_test)
y_predict[:10]

array([0, 0, 0, 0, 1, 0, 1, 0, 1, 0], dtype=int64)

The 'y_predict' array contains predictions that answer the question: "__Was a given passenger more likely than not to have survived the sinking of the Titanic?__"

If we had the actual __survived/perished__ indicators for the test data, we could now check the performance of the model via a variety of error metrics (e.g., accuracy, specificity, precision, recall, AUC, etc.).

### Case Study: Logistic Regression with scikit-learn

https://nbviewer.jupyter.org/gist/justmarkham/6d5c061ca5aee67c4316471f8c2ae976

# Module 7 Assignment Guidelines / Requirements
https://yu.instructure.com/courses/63488/assignments/324308?module_item_id=1157311